### ISSUE TITLE
Yip within Kairos fails to respect the shebang in a script

### ISSUE THREAD

**Issue by mauromorales** [kairos-io/kairos#2958]
Follow up from kairos-io/kairos-docs#289

Let's check if we can fix the problem, if so, we could remove the added documentation.

The documentation ticket also explains the problem further, but we think this is related to the console running the command:

> yes but both use different consoles to run no? https://github.com/kairos-io/kairos-agent/blob/f24511b2dd49876e05c16b33a22423df3f3b693f/pkg/cloudinit/console.go#L47 vs https://github.com/mudler/yip/blob/87b55bb1813f1f132a2a0f8242eadd9d702ce5f2/pkg/console/console.go#L56
> So maybe our runner does something weird underneath?

_Originally posted by @Itxaka in https://github.com/kairos-io/kairos-docs/pull/289#discussion_r1804398419_

**Comment by mauromorales** [comment]
@mudler-agent I believe the issue comes from teh underlying shell depending on the different linux distributions

**Comment by mudler-agent** [comment]
Confirmed. Both current paths run inline commands with `sh -c`:

- yip uses `exec.Command("sh", "-c", command)`.
- kairos-agent uses `runner.InitCmd("sh", "-c", command)`.

The runner does not change the shell. Each distribution resolves `sh` to its own implementation. Bash syntax such as `[[ ... ]]` can therefore work on one flavor and fail on another.

The original Ubuntu CLI result also hid the failure. Ubuntu resolves `/bin/sh` to `dash`. For `if [[ ... ]]; then ...; fi`, `dash` prints `[[: not found`, but the compound command exits with status 0. Yip only logs the captured output at debug level, so the normal output says that the stage completed.

An executable script path still honors its shebang. Inline commands have no shebang, so they must use POSIX `sh` syntax or invoke the required shell explicitly, for example `bash -c 'if [[ ... ]]; then ...; fi'`.\n\nI do not see a kairos-agent defect to patch here. Changing all inline commands to Bash would break shell portability and images without Bash. The existing documentation is the appropriate fix.

**Comment by ci-robbot** [comment]
Picked this up. The cause is narrower than the console difference we suspected, and both consoles behave the same here.

**What actually happens**

Both `yip`'s `StandardConsole` and kairos-agent's `cloudInitConsole` run a command as `sh -c <command>`. So the shebang only matters in one of the two shapes a command can take:

- `commands: [/path/to/script.sh]` with an executable script. Works today. `sh` forks and execve's the path, and the kernel reads the shebang.
- an inline block whose text starts with `#!`. Broken. The text goes to `sh -c`, which reads the shebang as a comment and runs the body itself.

The second shape is what produced the `[[: not found` in #2946:

```
$ sh -c "$(cat script.sh)"
sh: 2: [[: not found
```

It is quieter than that error suggests. With this config on current yip:

```yaml
stages:
  test:
    - name: "inline bash script"
      commands:
        - |
          #!/bin/bash
          if [[ "foo" == "foo" ]]; then echo ok > /tmp/marker.txt; fi
```

yip exits 0, reports the stage as done, and `/tmp/marker.txt` is never written. The `if` yields no error status under `sh`, so nothing surfaces.

This is not distribution dependent in the way we assumed. Any `/bin/sh` that is not bash drops the bash-only syntax, and a distribution where `/bin/sh` is bash hides the bug.

**Fix**

mudler/yip#333 writes a command that starts with `#!` to a private 0700 file and runs it by path, so the kernel starts the interpreter named on the first line. Commands without a shebang are untouched. It carries a unit spec and an end-to-end spec that runs `[[` through the real console and fails on master.

Once that lands and yip is bumped in kairos-agent, the documentation added in kairos-io/kairos-docs#289 can stay as the explicit `bash /path/to/script.sh` option, but it stops being the only way to get bash.

**Comment by ci-robbot** [comment]
This is unblocked, so I picked it up. PR: #4531.

Correcting my own note from 2026-09-02, which said the fix was still pending upstream: mudler/yip#333 merged later the same day.

**Nothing needed on the Kairos side beyond the bump.** Worth stating explicitly, because my earlier comment said both `StandardConsole` and the agent's `cloudInitConsole` run `sh -c` and left it open whether the agent needed its own fix. It does not. yip#333 landed in the `commands` plugin, above the `Console` interface, so the agent's console inherits the behaviour with no code change.

**The bump is unusually clean.** yip `master` is `v1.26.0` plus that single commit:

```
$ gh api /repos/mudler/yip/compare/v1.26.0...master --jq '.ahead_by'
1
```

So the PR moves `go.mod`, `go.sum`, and nothing else. No transitive drag.

**Red/green, since that is the part worth checking.** The specs go through the real yip executor and the agent's own console. On the current pin three fail, one of them with the reporter's shape of the failure:

```
sh: syntax error: unexpected "("
```

and all pass on the bumped pin. Full suite diffed against `master` in both trees, no regressions.

**One decision left for @mudler:** the pin is a pseudo-version, because `v1.26.0` was tagged 2026-08-26, a week before #333 merged. If you cut a yip release I will repoint the PR at the tag; the diff is otherwise identical. Happy either way, just did not want to sit on the fix waiting for a tag.

@mauromorales on your point about the underlying shell differing per distribution: that turned out not to be the cause. Both consoles use `sh -c` and behave the same. What varies per distribution is only whether `/bin/sh` happens to be bash, which hid the bug on some images rather than causing it.

**Comment by ci-robbot** [comment]
Picking this up for QA.

Testing the fix in #4531 (head `7134801b`). The earlier QA claim on that PR at 17:01Z never posted a result, and this issue was moved back into the QA column at 20:05Z, so treating it as an open QA request.

**Comment by ci-robbot** [comment]
### QA Result: PASS

| Check | Result |
| --- | --- |
| Environment | ISO built with `make iso` from #4531 head `7134801b` (`kairos v4.3.0-1-g7134801b9`, yip `v1.26.1-0.20260902132557-8b1bbf1c6b6e`). Live boot in QEMU/KVM under OVMF, 4 vCPU / 4G, cloud-config delivered as a NoCloud `cidata` cdrom. Guest `/bin/sh` is bash. |
| Reproduction | Four `boot`-stage commands from a real user cloud-config: an inline block starting with `#!/bin/bash`, a control block with no shebang, an inline block starting with `#!/usr/bin/awk -f`, and a leftover-file probe. Then a same-VM red/green: the PR-parent binary (`d17464a2`, yip `v1.26.0`) and the PR binary run the identical stage through `kairos-agent run-stage --override-cloud-init-paths`. |
| Expected | A command whose text starts with `#!` is run by the interpreter that line names. A command without a shebang is still handed to `sh -c` verbatim. No temp script is left behind. |
| Actual | All four as expected, and the parent binary fails the same checks in the same VM. |
| Proofs | [Full logs (gist)](https://gist.github.com/ci-robbot/d58051aa35b7a05dd573afe6402bdae2), console screendump below. |

**Boot stage, PR ISO.** This is the user-facing path, not a test harness:

```
=== agent version ===        kairos v4.3.0-1-g7134801b9
=== /bin/sh resolves to ===  /usr/bin/bash
A  #!/bin/bash      argv0=/tmp/yip-command-83f2579a3445429e
                    interpreter=/usr/bin/bash
                    double_bracket=ok
                    array_index_1=two
B  no shebang       argv0=sh
C  #!/usr/bin/awk   awk-interpreted-the-script
D  during the run   /tmp/yip-command-0a230550db2bcd07
   after the run    ls: cannot access '/tmp/yip-command-*': No such file or directory
```

`argv0` is the discriminator, because this image's `/bin/sh` **is** bash, so the reporter's `[[: not found` symptom is hidden here. Broken gives `argv0=sh`; fixed gives `argv0=/tmp/yip-command-*`. Probe C is the decisive one: `sh -c` cannot possibly run an awk `BEGIN` block, so the marker file only appears if the kernel read the shebang and execve'd awk.

**Same-VM red/green.** Both binaries built `CGO_ENABLED=0` from the two commits, copied into the booted guest, run against the same stage file:

```
################ old pin (d17464a2, yip v1.26.0) ################
ERR Error on file /tmp/rg/ci/stage.yaml on stage non shell interpreter awk: 1 error occurred:
        * failed to run #!/usr/bin/awk -f
WRN Some errors found but were ignored. Enable --strict mode to fail on those
bash     : argv0=sh double_bracket=ok array_index_1=two
awk      : MISSING
control  : argv0=sh

################ new pin (7134801b) ################
(no errors)
bash     : argv0=/tmp/yip-command-84422622768e009b double_bracket=ok array_index_1=two
awk      : awk-interpreted-the-script
control  : argv0=sh
```

Worth noting from the old-pin run: the awk step *did* log an error, but the stage still exited 0 and only warned. That matches the "quiet" failure mode described on the issue.

```
kairos-agent-old  dep github.com/mudler/yip v1.26.0
kairos-agent-new  dep github.com/mudler/yip v1.26.1-0.20260902132557-8b1bbf1c6b6e
```

**Regression checks.** `CGO_ENABLED=0 go test -count=1 ./agent/pkg/cloudinit/...` is green on the branch. The control probe confirms shebang-less commands are unchanged (`argv0=sh` on both pins). The diff against its own base `d17464a2` is the bump and the new spec file only, nothing else:

```
agent/pkg/cloudinit/shebang_test.go | 165 ++++++++++
go.mod                              |   2 +-
go.sum                              |   6 +-
```

so the go-containerregistry and quic-go churn is not part of this PR.

![console](https://raw.githubusercontent.com/ci-forks/kairos/qa-proofs-2958/console.png)

**Summary:** PASS. On the PR ISO an inline `boot`-stage command starting with `#!` is executed by the interpreter its shebang names, including a non-shell interpreter, the temp script is cleaned up afterwards, and commands without a shebang still go to `sh -c` unchanged. The PR parent fails the same checks in the same VM, so this is a real red/green and not a green-only guard. #2958 is resolved by #4531.

**Comment by ci-robbot** [comment]
Update: the fix is now on `master` under a real release tag, not the pseudo-version.

mudler/yip cut **v1.26.1** on 2026-09-09, and `kairos/go.mod` on `master` is already at `github.com/mudler/yip v1.26.1`. The tag contains exactly the commit my QA above ran against, so nothing about that verdict changes:

```
8b1bbf1c6b6e fix(commands): honour a shebang at the start of a command (#333)
fe549747402b fix(hetzner): read the native metadata API, keep provisioning SSH keys (#336)
```

That first line is this bug. (The second is #4374, which is already closed.) My QA PASS was against `v1.26.1-0.20260902132557-8b1bbf1c6b6e`, which is that same commit as a pseudo-version, so the shipped code is byte-identical to what was tested.

The behaviour is therefore fixed on `master` today. #4531 is now only the regression test for it, no dependency change left in the diff, which is why it can merge on its own schedule without holding this open.

Also worth deleting when someone touches it: kairos-docs still carries the workaround this issue was filed to remove, from kairos-io/kairos-docs#289.

**Comment by ci-robbot** [comment]
Docs follow-up opened as kairos-io/kairos-docs#690, which covers the second half of this issue ("if so, we could remove the added documentation").

It rewrites rather than removes: the interpreter-prefix form from kairos-io/kairos-docs#289 is still the only thing that works on v4.3.0 and earlier, so deleting it would strand those users. The shebang form now leads, with the version boundary called out.
